### ISSUE TITLE
[📍] Global mocks and dummy data

### ISSUE THREAD

**Issue by dave-2** [catapult-project/catapult#4495]
From a discussion with @anniesullie.

## Current state
Each unit test is responsible for its own mocks. Generally, unit tests for module `foo.py` mock out all calls made from `foo.py` that eventually reach an external service. E.g. `job_test.py` mocks out all calls to Monorail (for bug posting), but also all calls to `Commit.AsDict()` since that implicitly calls out to Gerrit.

## Problems
* It's hard to know which calls are safe and which require mocks.
* Some objects with multiple external dependencies (e.g. Job, Change) are hard to create in a test, since they require many mocks.
* Dependencies between modules are not well tested. E.g. `results2_test.py` mocks out all calls to `Job.AsDict()`, so if the return value of `AsDict()` changes, the tests still pass when they should fail.

## Proposed solution
Create a `TestCase` class that mocks and/or stubs out all external services, creating a "protective bubble" around all the tests. We can include some dummy data that all tests are free to rely on (e.g. repository url mappings, isolate hashes, some pre-fab jobs in various states). With this in place, all tests are free to instantiate and manipulate any object without needing to add additional test-specific mocks. We can also write full integration tests that need no additional mocks.

@anniesullie @perezju @simonhatch

**Comment by dave-2** [comment]
My original plan was to have a base test class that has all this dummy data already set up. But since Juan mentioned that it's valuable to keep cause and effect close together, I decided that we can create a set of stub helpers. Any unit test can import these helpers to get some pre-fab data of various states. For example, we might do something like:

```python
# job_state_test.py
def testDifference(self):
  change_test.SetMidpoint(self)

  quests = [quest_test.QuestStub(fails=change_test.CHANGE_B, passes=change_test.CHANGE_A)]
  s = job_state.JobState(quests, comparison_mode=job_state.FUNCTIONAL)
  s.AddChange(change_test.CHANGE_A)
  s.AddChange(change_test.CHANGE_B)

  s.ScheduleWork()
  s.Explore()

  # Assert that the midpoint change was added.
  expected = [change_test.CHANGE_A, change_test.CHANGE_B, change_test.CHANGE_MIDPOINT]
  s.assertEqual(state._changes, expected)
```

Many of the individual unit tests already create this kind of pre-fab data for their own use. So we just need to formalize them as APIs that can be used by other unit tests.

**Comment by perezju** [comment]
Yeah, in the current direction that you are thinking, I guess the best solution is to provide custom "stubs" or "factories" that tests can use to create mock objects with sensible default behaviors, but that tests can also customize if they want to test against a particular mock output or behavior.
